File: app/src/main/java/me/taosunkist/hello/ui/frgment/countdowntimer/squareprogressbar/SquareProgressView.java
package me.taosunkist.hello.ui.frgment.countdowntimer.squareprogressbar;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.CornerPathEffect;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Paint.Style;
import android.graphics.Path;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import java.text.DecimalFormat;

import me.taosunkist.hello.Dimens;

public class SquareProgressView extends View {

    private float progress;
    private Paint progressBarPaint;
    private Paint outlinePaint;
    private Paint textPaint;

    private float widthInDp = 10;
    private float strokewidth = 0;
    private Canvas canvas;

    private boolean outline = false;
    private boolean startline = false;
    private boolean showProgress = false;
    private boolean centerline = false;

    private boolean roundedCorners = false;
    private float roundedCornersRadius = 10;

    private PercentStyle percentSettings = new PercentStyle(Align.CENTER, 150, true);
    private boolean clearOnHundred = false;
    private boolean isIndeterminate = false;
    private int indeterminate_count = 1;

    private float indeterminate_width = 20.0f;

    public SquareProgressView(Context context) {
        super(context);
        initializePaints(context);
    }

    public SquareProgressView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initializePaints(context);
    }

    public SquareProgressView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initializePaints(context);
    }

    private void initializePaints(Context context) {
        progressBarPaint = new Paint();
        progressBarPaint.setColor(context.getResources().getColor(android.R.color.holo_green_dark));
        progressBarPaint.setStrokeWidth(Dimens.INSTANCE.dpToPx((int) widthInDp));
        progressBarPaint.setAntiAlias(true);
        progressBarPaint.setStyle(Style.STROKE);

        outlinePaint = new Paint();
        outlinePaint.setColor(context.getResources().getColor(android.R.color.black));
        outlinePaint.setStrokeWidth(1);
        outlinePaint.setAntiAlias(true);
        outlinePaint.setStyle(Style.STROKE);

        textPaint = new Paint();
        textPaint.setColor(context.getResources().getColor(android.R.color.black));
        textPaint.setAntiAlias(true);
        textPaint.setStyle(Style.STROKE);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        this.canvas = canvas;
        super.onDraw(canvas);
        strokewidth = Dimens.INSTANCE.dpToPx((int) widthInDp);
        int cW = getWidth();
        int cH = getHeight();
        Log.i("taohui", "" + cW + "-" + cH + ", " + Dimens.INSTANCE.getScreenWidth() + "-" + Dimens.INSTANCE.getScreenHeight());
        float scope = (2 * cW) + (2 * cH) - (4 * strokewidth);
        float hSw = strokewidth / 2;

        if (isOutline()) {
            drawOutline();
        }

        if (isStartline()) {
            drawStartline();
        }

        if (isShowProgress()) {
            drawPercent(percentSettings);
        }

        if (isCenterline()) {
            drawCenterline(strokewidth);
        }

        if ((isClearOnHundred() && progress == 100.0) || (progress <= 0.0)) {
            return;
        }

        if (isIndeterminate()) {
            Path path = new Path();
            DrawStop drawEnd = getDrawEnd((scope / 100) * Float.valueOf(String.valueOf(indeterminate_count)), canvas);

            if (drawEnd.place == Place.TOP) {
                path.moveTo(drawEnd.location - indeterminate_width - strokewidth, hSw);
                path.lineTo(drawEnd.location, hSw);
                canvas.drawPath(path, progressBarPaint);
            }

            if (drawEnd.place == Place.RIGHT) {
                path.moveTo(cW - hSw, drawEnd.location - indeterminate_width);
                path.lineTo(cW - hSw, strokewidth + drawEnd.location);
                canvas.drawPath(path, progressBarPaint);
            }

            if (drawEnd.place == Place.BOTTOM) {
                path.moveTo(drawEnd.location - indeterminate_width - strokewidth, cH - hSw);
                path.lineTo(drawEnd.location, cH - hSw);
                canvas.drawPath(path, progressBarPaint);
            }

            if (drawEnd.place == Place.LEFT) {
                path.moveTo(hSw, drawEnd.location - indeterminate_width - strokewidth);
                path.lineTo(hSw, drawEnd.location);
                canvas.drawPath(path, progressBarPaint);
            }

            indeterminate_count++;
            if (indeterminate_count > 100) {
                indeterminate_count = 0;
            }
            invalidate();
        } else {
            Path path = new Path();
            DrawStop drawEnd = getDrawEnd((scope / 100) * Float.valueOf(String.valueOf(progress)), canvas);

            if (drawEnd.place == Place.TOP) {
                if (drawEnd.location > (cW / 2) && progress < 100.0) {
                    path.moveTo(cW / 2, hSw);
                    path.lineTo(drawEnd.location, hSw);
                } else {
                    path.moveTo(cW / 2, hSw);
                    path.lineTo(cW - hSw, hSw);
                    path.lineTo(cW - hSw, cH - hSw);
                    path.lineTo(hSw, cH - hSw);
                    path.lineTo(hSw, hSw);
                    path.lineTo(strokewidth, hSw);
                    path.lineTo(drawEnd.location, hSw);
                }
                canvas.drawPath(path, progressBarPaint);
            }

            if (drawEnd.place == Place.RIGHT) {
                path.moveTo(cW / 2, hSw);
                path.lineTo(cW - hSw, hSw);
                path.lineTo(cW - hSw, 0 + drawEnd.location);
                canvas.drawPath(path, progressBarPaint);
            }

            if (drawEnd.place == Place.BOTTOM) {
                path.moveTo(cW / 2, hSw);
                path.lineTo(cW - hSw, hSw);
                path.lineTo(cW - hSw, cH - hSw);
                path.lineTo(cW - strokewidth, cH - hSw);
                path.lineTo(drawEnd.location, cH - hSw);
                canvas.drawPath(path, progressBarPaint);
            }

            if (drawEnd.place == Place.LEFT) {
                path.moveTo(cW / 2, hSw);
                path.lineTo(cW - hSw, hSw);
                path.lineTo(cW - hSw, cH - hSw);
                path.lineTo(hSw, cH - hSw);
                path.lineTo(hSw, cH - strokewidth);
                path.lineTo(hSw, drawEnd.location);
                canvas.drawPath(path, progressBarPaint);
            }
        }
    }

    private void drawStartline() {
        Path outlinePath = new Path();
        outlinePath.moveTo(getWidth() / 2, 0);
        outlinePath.lineTo(getWidth() / 2, strokewidth);
        canvas.drawPath(outlinePath, outlinePaint);
    }

    private void drawOutline() {
        Path outlinePath = new Path();
        outlinePath.moveTo(0, 0);
        outlinePath.lineTo(getWidth(), 0);
        outlinePath.lineTo(getWidth(), getHeight());
        outlinePath.lineTo(0, getHeight());
        outlinePath.lineTo(0, 0);
        canvas.drawPath(outlinePath, outlinePaint);
    }

    public float getProgress() {
        return progress;
    }

    public void setProgress(float progress) {
        this.progress = progress;
        this.invalidate();
    }

    public void setColor(int color) {
        progressBarPaint.setColor(color);
        this.invalidate();
    }

    public void setWidthInDp(int width) {
        this.widthInDp = width;
        progressBarPaint.setStrokeWidth(Dimens.INSTANCE.dpToPx((int) widthInDp));
        this.invalidate();
    }

    public boolean isOutline() {
        return outline;
    }

    public void setOutline(boolean outline) {
        this.outline = outline;
        this.invalidate();
    }

    public boolean isStartline() {
        return startline;
    }

    public void setStartline(boolean startline) {
        this.startline = startline;
        this.invalidate();
    }

    private void drawPercent(PercentStyle setting) {
        textPaint.setTextAlign(setting.getAlign());
        if (setting.getTextSize() == 0) {
            textPaint.setTextSize((getHeight() / 10) * 4);
        } else {
            textPaint.setTextSize(setting.getTextSize());
        }

        String percentString = new DecimalFormat("###").format(getProgress());
        if (setting.isPercentSign()) {
            percentString = percentString + percentSettings.getCustomText();
        }

        textPaint.setColor(percentSettings.getTextColor());

        canvas.drawText(percentString, getWidth() / 2, (int) ((getHeight() / 2) - ((textPaint.descent() + textPaint.ascent()) / 2)), textPaint);
    }

    public boolean isShowProgress() {
        return showProgress;
    }

    public void setShowProgress(boolean showProgress) {
        this.showProgress = showProgress;
        this.invalidate();
    }

    public void setPercentStyle(PercentStyle percentSettings) {
        this.percentSettings = percentSettings;
        this.invalidate();
    }

    public PercentStyle getPercentStyle() {
        return percentSettings;
    }

    public void setClearOnHundred(boolean clearOnHundred) {
        this.clearOnHundred = clearOnHundred;
        this.invalidate();
    }

    public boolean isClearOnHundred() {
        return clearOnHundred;
    }

    private void drawCenterline(float strokewidth) {
        float centerOfStrokeWidth = strokewidth / 2;
        Path centerlinePath = new Path();
        centerlinePath.moveTo(centerOfStrokeWidth, centerOfStrokeWidth);
        centerlinePath.lineTo(getWidth() - centerOfStrokeWidth, centerOfStrokeWidth);
        centerlinePath.lineTo(getWidth() - centerOfStrokeWidth, getHeight() - centerOfStrokeWidth);
        centerlinePath.lineTo(centerOfStrokeWidth, getHeight() - centerOfStrokeWidth);
        centerlinePath.lineTo(centerOfStrokeWidth, centerOfStrokeWidth);
        canvas.drawPath(centerlinePath, outlinePaint);
    }

    public boolean isCenterline() {
        return centerline;
    }

    public void setCenterline(boolean centerline) {
        this.centerline = centerline;
        this.invalidate();
    }

    public boolean isIndeterminate() {
        return isIndeterminate;
    }

    public void setIndeterminate(boolean indeterminate) {
        isIndeterminate = indeterminate;
        this.invalidate();
    }

    public DrawStop getDrawEnd(float percent, Canvas canvas) {
        DrawStop drawStop = new DrawStop();
        strokewidth = Dimens.INSTANCE.dpToPx((int) widthInDp);
        float halfOfTheImage = getWidth() / 2;

        // top right
        if (percent > halfOfTheImage) {
            float second = percent - (halfOfTheImage);

            // right
            if (second > (getHeight() - strokewidth)) {
                float third = second - (getHeight() - strokewidth);

                // bottom
                if (third > (getWidth() - strokewidth)) {
                    float forth = third - (getWidth() - strokewidth);

                    // left
                    if (forth > (getHeight() - strokewidth)) {
                        float fifth = forth - (getHeight() - strokewidth);

                        // top left
                        if (fifth == halfOfTheImage) {
                            drawStop.place = Place.TOP;
                            drawStop.location = halfOfTheImage;
                        } else {
                            drawStop.place = Place.TOP;
                            drawStop.location = strokewidth + fifth;
                        }
                    } else {
                        drawStop.place = Place.LEFT;
                        drawStop.location = getHeight() - strokewidth - forth;
                    }

                } else {
                    drawStop.place = Place.BOTTOM;
                    drawStop.location = getWidth() - strokewidth - third;
                }
            } else {
                drawStop.place = Place.RIGHT;
                drawStop.location = strokewidth + second;
            }

        } else {
            drawStop.place = Place.TOP;
            drawStop.location = halfOfTheImage + percent;
        }

        return drawStop;
    }

    public void setRoundedCorners(boolean roundedCorners, float radius) {
        this.roundedCorners = roundedCorners;
        this.roundedCornersRadius = radius;
        if (roundedCorners) {
            progressBarPaint.setPathEffect(new CornerPathEffect(roundedCornersRadius));
        } else {
            progressBarPaint.setPathEffect(null);
        }
        this.invalidate();
    }

    public boolean isRoundedCorners() {
        return roundedCorners;
    }

    private class DrawStop {

        private Place place;
        private float location;

        public DrawStop() {

        }
    }

    public enum Place {
        TOP, RIGHT, BOTTOM, LEFT
    }
}

File: app/src/main/java/me/taosunkist/hello/ui/reusable/SquareProgressView.kt
package me.taosunkist.hello.ui.reusable

import android.R
import android.content.Context
import android.graphics.Canvas
import android.graphics.CornerPathEffect
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.view.View
import androidx.annotation.ColorInt
import me.taosunkist.hello.utility.Dimens.dpToPx
import java.text.DecimalFormat

class SquareProgressView : View {
    private var progress = 0f
    private var progressBarPaint: Paint = Paint()
    private var outlinePaint: Paint = Paint()
    private var textPaint: Paint = Paint()
    private var widthInDp = 10f
    private var strokewidth = 0f
    val path = Path()

    var isOutline = false
        set(outline) {
            field = outline
            this.invalidate()
        }
    var isStartline = false
        set(startline) {
            field = startline
            this.invalidate()
        }
    var isShowProgress = false
        set(showProgress) {
            field = showProgress
            this.invalidate()
        }
    var isCenterline = false
        set(centerline) {
            field = centerline
            this.invalidate()
        }
    var isRoundedCorners = false
        private set
    private var roundedCornersRadius = 10f
    private var percentSettings = PercentStyle(Paint.Align.CENTER, 150f, true)
    var isClearOnHundred = false
        set(clearOnHundred) {
            field = clearOnHundred
            this.invalidate()
        }
    var isIndeterminate = false
        set(indeterminate) {
            field = indeterminate
            this.invalidate()
        }
    private var indeterminateCount = 1
    private val indeterminateWidth = 20.0f

    constructor(context: Context) : super(context) {
        initializePaints(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        initializePaints(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        initializePaints(context)
    }

    private fun initializePaints(context: Context) {
        progressBarPaint.color = context.resources.getColor(R.color.holo_green_dark)
        progressBarPaint.strokeWidth = dpToPx(widthInDp.toInt()).toFloat()
        progressBarPaint.isAntiAlias = true
        progressBarPaint.style = Paint.Style.STROKE
        outlinePaint = Paint()
        outlinePaint.color = context.resources.getColor(R.color.black)
        outlinePaint.strokeWidth = 1f
        outlinePaint.isAntiAlias = true
        outlinePaint.style = Paint.Style.STROKE
        textPaint = Paint()
        textPaint.color = context.resources.getColor(R.color.black)
        textPaint.isAntiAlias = true
        textPaint.style = Paint.Style.STROKE
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        strokewidth = dpToPx(widthInDp.toInt()).toFloat()
        val cW = width
        val cH = height
        val scope = 2 * cW + 2 * cH - 4 * strokewidth
        val hSw = strokewidth / 2
        if (isOutline) {
            drawOutline(canvas)
        }
        if (isStartline) {
            drawStartline(canvas)
        }
        if (isShowProgress) {
            drawPercent(canvas, percentSettings)
        }
        if (isCenterline) {
            drawCenterLine(canvas, strokewidth)
        }
        if (isClearOnHundred && progress.toDouble() == 100.0 || progress <= 0.0) {
            return
        }
        if (isIndeterminate) {
            path.reset()
            val drawEnd = getDrawEnd(scope / 100 * java.lang.Float.valueOf(indeterminateCount.toString()), canvas)
            if (drawEnd.place == Place.TOP) {
                path.moveTo(drawEnd.location - indeterminateWidth - strokewidth, hSw)
                path.lineTo(drawEnd.location, hSw)
                canvas.drawPath(path, progressBarPaint)
            }
            if (drawEnd.place == Place.RIGHT) {
                path.moveTo(cW - hSw, drawEnd.location - indeterminateWidth)
                path.lineTo(cW - hSw, strokewidth + drawEnd.location)
                canvas.drawPath(path, progressBarPaint)
            }
            if (drawEnd.place == Place.BOTTOM) {
                path.moveTo(drawEnd.location - indeterminateWidth - strokewidth, cH - hSw)
                path.lineTo(drawEnd.location, cH - hSw)
                canvas.drawPath(path, progressBarPaint)
            }
            if (drawEnd.place == Place.LEFT) {
                path.moveTo(hSw, drawEnd.location - indeterminateWidth - strokewidth)
                path.lineTo(hSw, drawEnd.location)
                canvas.drawPath(path, progressBarPaint)
            }
            indeterminateCount++
            if (indeterminateCount > 100) {
                indeterminateCount = 0
            }
            invalidate()
        } else {
            path.reset()
            val drawEnd = getDrawEnd(scope / 100 * java.lang.Float.valueOf(progress.toString()), canvas)
            if (drawEnd.place == Place.TOP) {
                if (drawEnd.location > cW / 2 && progress < 100.0) {
                    path.moveTo((cW / 2).toFloat(), hSw)
                    path.lineTo(drawEnd.location, hSw)
                } else {
                    path.moveTo((cW / 2).toFloat(), hSw)
                    path.lineTo(cW - hSw, hSw)
                    path.lineTo(cW - hSw, cH - hSw)
                    path.lineTo(hSw, cH - hSw)
                    path.lineTo(hSw, hSw)
                    path.lineTo(strokewidth, hSw)
                    path.lineTo(drawEnd.location, hSw)
                }
                canvas.drawPath(path, progressBarPaint)
            }
            if (drawEnd.place == Place.RIGHT) {
                path.moveTo((cW / 2).toFloat(), hSw)
                path.lineTo(cW - hSw, hSw)
                path.lineTo(cW - hSw, 0 + drawEnd.location)
                canvas.drawPath(path, progressBarPaint)
            }
            if (drawEnd.place == Place.BOTTOM) {
                path.moveTo((cW / 2).toFloat(), hSw)
                path.lineTo(cW - hSw, hSw)
                path.lineTo(cW - hSw, cH - hSw)
                path.lineTo(cW - strokewidth, cH - hSw)
                path.lineTo(drawEnd.location, cH - hSw)
                canvas.drawPath(path, progressBarPaint)
            }
            if (drawEnd.place == Place.LEFT) {
                path.moveTo((cW / 2).toFloat(), hSw)
                path.lineTo(cW - hSw, hSw)
                path.lineTo(cW - hSw, cH - hSw)
                path.lineTo(hSw, cH - hSw)
                path.lineTo(hSw, cH - strokewidth)
                path.lineTo(hSw, drawEnd.location)
                canvas.drawPath(path, progressBarPaint)
            }
        }
    }

    private fun drawStartline(canvas: Canvas) {
        val outlinePath = Path()
        outlinePath.moveTo((width / 2).toFloat(), 0f)
        outlinePath.lineTo((width / 2).toFloat(), strokewidth)
        canvas.drawPath(outlinePath, outlinePaint)
    }

    private fun drawOutline(canvas: Canvas) {
        val outlinePath = Path()
        outlinePath.moveTo(0f, 0f)
        outlinePath.lineTo(width.toFloat(), 0f)
        outlinePath.lineTo(width.toFloat(), height.toFloat())
        outlinePath.lineTo(0f, height.toFloat())
        outlinePath.lineTo(0f, 0f)
        outlinePaint.style = Paint.Style.FILL
        canvas.drawPath(outlinePath, outlinePaint)
    }

    fun getProgress(): Float {
        return progress
    }

    fun setProgress(progress: Float) {
        this.progress = progress
        this.invalidate()
    }

    fun setColor(color: Int) {
        progressBarPaint.color = color
        this.invalidate()
    }

    fun setWidthInDp(width: Int) {
        widthInDp = width.toFloat()
        progressBarPaint.strokeWidth = dpToPx(widthInDp.toInt()).toFloat()
        this.invalidate()
    }

    private fun drawPercent(canvas: Canvas, setting: PercentStyle) {
        textPaint.textAlign = setting.align
        if (setting.textSize == 0f) {
            textPaint.textSize = (height / 10 * 4).toFloat()
        } else {
            textPaint.textSize = setting.textSize
        }
        var percentString = DecimalFormat("###").format(getProgress().toDouble())
        if (setting.isPercentSign) {
            percentString += percentSettings.customText
        }
        textPaint.color = percentSettings.textColor
        canvas.drawText(percentString,
            (width / 2).toFloat(),
            (height / 2 - (textPaint.descent() + textPaint.ascent()) / 2),
            textPaint)
    }

    var percentStyle: PercentStyle
        get() = percentSettings
        set(percentSettings) {
            this.percentSettings = percentSettings
            this.invalidate()
        }

    private fun drawCenterLine(canvas: Canvas, strokeWidth: Float) {
        val centerOfStrokeWidth = strokeWidth / 2
        val centerLinePath = Path()
        centerLinePath.moveTo(centerOfStrokeWidth, centerOfStrokeWidth)
        centerLinePath.lineTo(width - centerOfStrokeWidth, centerOfStrokeWidth)
        centerLinePath.lineTo(width - centerOfStrokeWidth, height - centerOfStrokeWidth)
        centerLinePath.lineTo(centerOfStrokeWidth, height - centerOfStrokeWidth)
        centerLinePath.lineTo(centerOfStrokeWidth, centerOfStrokeWidth)
        canvas.drawPath(centerLinePath, outlinePaint)
    }

    private fun getDrawEnd(percent: Float, canvas: Canvas?): DrawStop {
        val drawStop = DrawStop()
        strokewidth = dpToPx(widthInDp.toInt()).toFloat()
        val halfOfTheImage = (width / 2).toFloat()

        // top right
        if (percent > halfOfTheImage) {
            val second = percent - halfOfTheImage

            // right
            if (second > height - strokewidth) {
                val third = second - (height - strokewidth)

                // bottom
                if (third > width - strokewidth) {
                    val forth = third - (width - strokewidth)

                    // left
                    if (forth > height - strokewidth) {
                        val fifth = forth - (height - strokewidth)

                        // top left
                        if (fifth == halfOfTheImage) {
                            drawStop.place = Place.TOP
                            drawStop.location = halfOfTheImage
                        } else {
                            drawStop.place = Place.TOP
                            drawStop.location = strokewidth + fifth
                        }
                    } else {
                        drawStop.place = Place.LEFT
                        drawStop.location = height - strokewidth - forth
                    }
                } else {
                    drawStop.place = Place.BOTTOM
                    drawStop.location = width - strokewidth - third
                }
            } else {
                drawStop.place = Place.RIGHT
                drawStop.location = strokewidth + second
            }
        } else {
            drawStop.place = Place.TOP
            drawStop.location = halfOfTheImage + percent
        }
        return drawStop
    }

    fun setRoundedCorners(roundedCorners: Boolean, radius: Float) {
        isRoundedCorners = roundedCorners
        roundedCornersRadius = radius
        if (roundedCorners) {
            progressBarPaint.pathEffect = CornerPathEffect(roundedCornersRadius)
        } else {
            progressBarPaint.pathEffect = null
        }
        this.invalidate()
    }

    inner class DrawStop {
        var place: Place? = null
        var location = 0f
    }

    enum class Place {
        TOP, RIGHT, BOTTOM, LEFT
    }

    fun setProgressBarColor(@ColorInt color: Int) {
        progressBarPaint.color = color
    }
}